.gitlab-ci.yml

```diff
@@ -58,37 +100,28 @@ build:native-deps:
 
 # stage deploy
 
+deploy:sls:master:
+  <<: *tmpl_master_rules
+  <<: *tmpl_deploy
+  environment:
+    name: production
+    url: https://$APP_URL
+
 deploy:sls:branch:
-  artifacts:
-    reports:
-      dotenv: deploy.env
+  <<: *tmpl_branch_rules
+  <<: *tmpl_deploy
   environment:
     name: review/$CI_COMMIT_REF_SLUG
     url: https://$APP_URL
     on_stop: deploy:sls:branch_remove
-  script:
-    - yarn pnpify serverless deploy --verbose
-    - echo "APP_URL=$(yarn pnpify serverless info --verbose | grep -w "CloudFrontDistributionDomain" | cut -d ' ' -f2)" >> deploy.env
-  only:
-    - external_pull_requests
-  needs:
-    - build
-    - build:native-deps
-  stage: deploy
-  variables:
-    SLS_STAGE: $CI_COMMIT_REF_SLUG
 
 deploy:sls:branch_remove:
+  <<: *tmpl_branch_remove_rules
+  dependencies:
+    - build:native-deps
   environment:
     name: review/$CI_COMMIT_REF_SLUG
     action: stop
   script:
+    - git checkout $CI_COMMIT_SHA # the branch may not exist
```

Comment:
Isn't it gonna fail even before this in that case? Because checking out the branch is like a pre script step right?

Reply:
This job should execute after a pull request is merged and branch removed. 
The remove job doesn't rely on the branch name due to that, having a detached state with the git SHA is a better approach to avoid failure caused by removing the branch.

Reply:
"Dependencies" refer to the artifacts copied to the working directory, "needs" allow for "parallelization". For deploy, we "need" every previous stage to finish.

Reply:
I understand what you're trying to do, I'm just not convinced it works - if it usually tries to check out the branch before running the scripts, doesn't it already fail then before even getting to this script?

Reply:
I guess we can also just see if it works with this branch after merge? :sweat_smile:

Reply:
We can iterate on this later then 🙂